core/src/main/java/com/linecorp/armeria/server/http/MaxConnectionHandler.java

```diff
@@ -0,0 +1,72 @@
+/*
+ * Copyright 2017 LINE Corporation
+ *
+ * LINE Corporation licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+
+package com.linecorp.armeria.server.http;
+
+import java.util.concurrent.atomic.AtomicInteger;
+
+import io.netty.channel.ChannelHandler.Sharable;
+import io.netty.channel.ChannelHandlerContext;
+import io.netty.channel.ChannelInboundHandlerAdapter;
+
+@Sharable
+public final class MaxConnectionHandler extends ChannelInboundHandlerAdapter {
+
+    private final int maxConnections;
+    private final AtomicInteger connections = new AtomicInteger(0);
+
+    public MaxConnectionHandler(int maxConnections) {
+        this.maxConnections = validateMaxConnections(maxConnections);
+    }
+
+    @Override
+    public void channelActive(ChannelHandlerContext ctx) throws Exception {
+        int conn = connections.incrementAndGet();
+        if (conn > 0 && conn <= maxConnections) {
+            ctx.channel().closeFuture().addListener(future -> connections.decrementAndGet());
+
+            // Fire channelActive to the next handlers.
+            super.channelActive(ctx);
+        } else {
+            connections.decrementAndGet();
+            ctx.close();
+        }
+    }
+
+    /**
+     * Returns the maximum allowed number of open connections.
+     */
+    public int maxConnections() {
+        return maxConnections;
+    }
+
+    /**
+     * Returns the number of open connections.
+     */
+    public int currentConnections() {
```

Comment:
- How about `numConnections` which is shorter?
- This makes me think `maxConnections` could be renamed to `maxNumConnections` as well. WDYT?